blockchain/state_manager/src/lib.rs

```diff
@@ -223,10 +267,225 @@ where
+            .map_err(|err| {
+                Error::Other(format!(
+                    "failed to load tipset during msg wait searchback: {:}",
+                    err
+                ))
+            })?;
+        let cid = message.to_cid()?;
+        let r = self.tipset_executed_message(&tipset, &cid, message)?;
+
+        if let Some(receipt) = r {
+            return Ok(Some((tipset, receipt)));
+        }
+        self.search_back_for_message(&tipset, message)
+    }
+    /// returns a message receipt from a given tipset and message cid
+    pub fn get_receipt(&self, tipset: &Tipset, msg: &Cid) -> Result<MessageReceipt, Error> {
+        let m = chain::get_chain_message(self.get_block_store_ref(), msg)
+            .map_err(|e| Error::Other(e.to_string()))?;
+        let message_receipt = self.tipset_executed_message(tipset, msg, &*m)?;
+
+        if let Some(receipt) = message_receipt {
+            return Ok(receipt);
+        }
+
+        let maybe_tuple = self.search_back_for_message(tipset, &*m)?;
+        let message_receipt = maybe_tuple
+            .ok_or_else(|| {
+                Error::Other("Could not get receipt from search back message".to_string())
+            })?
+            .1;
+        Ok(message_receipt)
+    }
+
+    fn tipset_executed_message(
+        &self,
+        tipset: &Tipset,
+        cid: &Cid,
+        message: &dyn Message,
+    ) -> Result<Option<MessageReceipt>, Error> {
+        if tipset.epoch() == 0 {
+            return Ok(None);
+        }
+        let tipset = chain::tipset_from_keys(self.get_block_store_ref(), tipset.parents())
+            .map_err(|err| Error::Other(err.to_string()))?;
+        let messages = chain::messages_for_tipset(self.get_block_store_ref(), &tipset)
+            .map_err(|err| Error::Other(err.to_string()))?;
+        messages
+            .iter()
+            .enumerate()
+            .rev()
+            .filter(|(_, s)| s.from() == message.from())
+            .filter_map(|(index,s)| {
+                if s.sequence() == message.sequence() {
+                    if s.to_cid().map(|s| &s == cid).unwrap_or_default() {
+                        return Some(
+                            chain::get_parent_reciept(
+                                self.get_block_store_ref(),
+                                tipset.blocks().first().unwrap(),
+                                index as u64,
+                            )
+                            .map_err(|err| {
+                                Error::Other(err.to_string())
+                            }),
+                        );
+                    }
+                    if let Ok(m_cid) = s.to_cid()
+                    {
+                        let error_msg = format!("found message with equal nonce as the one we are looking for (F:{:} n {:}, TS: {:} n{:})",cid,message.sequence(),m_cid,s.sequence());
+                        return Some(Err(Error::Other(error_msg)))
+                    }
```

Comment:
why do you need this?
```suggestion
```

Reply:
For diagnostic purposes. if possible I want to be able to show the sequence on the cid but if not it should default. I could probably just return an error string based if the cid was found but the scope was so small that it didnt hurt

Reply:
I don't think it's worth it to have two seperate error returns just to have a slightly different error message, why don't you just debug print the cid? (also why are you putting `{:}` in the format? I can't find what that does, it's not in the std::fmt docs)

Reply:
I thought about it and debug_printing the cid will have less context just alone as compared to if it is printed with everything else. I will remove it but just felt really important. 

This is just calling on display {:} instead of fmt

Reply:
well I mean debug printing the option, (wouldn't that still display print the cid inside it?)

Also display is generally `{}`, is this for sure functional? (I have just never seen before)

Reply:
Yeah it is functional essentially just the same as { } https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=26ab59f94e2f4506c80a090e235685c7